R/to_labelled.R

```diff
@@ -206,21 +210,43 @@ memisc_to_labelled <- function(x) {
 #' identical(s1, to_labelled(f1))
 #' to_labelled(f1, labels)
 #' identical(s1, to_labelled(f1, labels))
+#'
+#' l <- labelled(c(1, 1, 2, 2, 9, 2, 1, 9), c("yes" = 1, "no" = 2, "don't know" = 9))
+#' f <- to_factor(l, levels = "p")
+#' f
+#' to_labelled(f)
+#' identical(to_labelled(f), l)
 to_labelled.factor <- function(x, labels = NULL, ...) {
+  vl <- var_label(x)
   if (is.null(labels)) {
-    vl <- var_label(x)
-    labs <- 1:length(levels(x))
-    names(labs) <- levels(x)
-    x <- labelled(as.numeric(x), labs)
-    var_label(x) <- vl
+    # check if levels are formatted as "[code] label"
+    l <- .get_prefixes.factor(x)
+    if (any(is.na(l$code)) | any(is.na(l$code))) {
+      # normal case
+      labs <- 1:length(levels(x))
+      names(labs) <- levels(x)
+      x <- labelled(as.numeric(x), labs)
+    } else {
+      # "[code] label" case
+      if (all(!is.na(suppressWarnings(as.numeric(l$code)))))
+        l$code <- as.numeric(l$code)
+        r <- l$levels
+        names(r) <- l$code
+        x <- forcats::fct_recode(x, !!!r) %>%
+          as.character()
```

Comment:
```suggestion
        levels(x) <- l$code
        x <- as.character(x)
```
That would avoid the new dependency to forcats

Reply:
mais bien sûr !

simple and elegant. I knew I was missing something here